### PR TITLE
regions plugin: snap to grid options

### PR DESCRIPTION
**Pull request by moritzvieli**
### Short description of changes:
There are use cases, where I want my regions to be able to snap to a grid. E.g. when I am editing a file with a fixed time signature. This PR introduces two new parameters for the regions plugin: "snapToGridInterval" and "snapToGridOffset". They both behave the same as the parameters "timeInterval" and "offset" in the timeline plugin.

### Breaking in the external API:
None

### Breaking changes in the internal API:
None

### Todos/Notes:
I don't know, whether it is a good idea to add the new parameters to the example, because it makes probably no sense there. But I can do it, if you want.

### Related Issues and other PRs:
None

### CODE REVIEW

File: src/plugin/regions.js [hunk 1]
```diff
@@ -816,9 +825,16 @@ export default class RegionsPlugin {
             }
 
             const end = this.wavesurfer.drawer.handleEvent(e);
+            const startUpdate = this.wavesurfer.regions.util.getRegionSnapToGridValue(
+                start * duration
+            );
+            const endUpdate = this.wavesurfer.regions.util.getRegionSnapToGridValue(
+                end * duration
+            );
+            console.log(startUpdate);
```
Comment (thijstriemstra) [inline src/plugin/regions.js]: no console statements

File: src/plugin/regions.js [hunk 2]
```diff
@@ -867,4 +883,25 @@ export default class RegionsPlugin {
 
         return min;
     }
+
+    /**
+     * Get the snapped to grid value.
+     *
+     * @param {*} value the value to snap to the grid, if needed
+     * @param {*} params the regions plugin params
+     */
```
Comment (thijstriemstra) [inline src/plugin/regions.js]: im sure there's a better type than `*` to document for these params.

**Comment by thijstriemstra** [comment]
And can you also this to the changelog?

File: src/plugin/regions.js (hunk 2 above)
Comment (moritzvieli) [reply src/plugin/regions.js]: Like this?

**Comment by moritzvieli** [comment]
> And can you also this to the changelog?

@thijstriemstra Thanks a lot for your input! I fixed the mentioned things. Please let me know, if there is anything else required.

File: CHANGES.md [hunk 3]
```diff
@@ -17,8 +17,9 @@ wavesurfer.js changelog
   - add `followCursorY` option (#1605)
 - Spectrogram plugin: fix `ready` listener when loading multiple
   audio files (#1572)
-- Regions plugin: fix `ready` listener when loading multiple
-  audio files (#1602)
+- Regions plugin:
+  - fix `ready` listener when loading multiple audio files (#1602)
+  - add `snapToGridInterval` and `snapToGridOffset` options
```
Comment (thijstriemstra) [inline CHANGES.md]: include PR nr

File: src/plugin/regions.js [hunk 4]
```diff
@@ -319,8 +319,9 @@ class Region {
                     );
 
                     // Update time
-                    const time =
-                        this.wavesurfer.drawer.handleEvent(e) * duration;
+                    const time = this.wavesurfer.regions.util.getRegionSnapToGridValue(
```
Comment (thijstriemstra) [inline src/plugin/regions.js]: add comment what value `time` is expected to be

File: src/plugin/regions.js [hunk 5]
```diff
@@ -342,8 +343,9 @@ class Region {
                         : null;
 
                     e.stopPropagation();
-                    startTime =
-                        this.wavesurfer.drawer.handleEvent(e, true) * duration;
+                    startTime = this.wavesurfer.regions.util.getRegionSnapToGridValue(
```
Comment (thijstriemstra) [inline src/plugin/regions.js]: add comment what value `startTime` is expected to be

File: src/plugin/regions.js [hunk 6]
```diff
@@ -622,6 +628,9 @@ export default class RegionsPlugin {
         this.params = params;
         this.wavesurfer = ws;
         this.util = ws.util;
+        this.util.getRegionSnapToGridValue = value => {
```
Comment (thijstriemstra) [inline src/plugin/regions.js]: why put this in `util`?

File: src/plugin/regions.js [hunk 7]
```diff
@@ -867,4 +882,25 @@ export default class RegionsPlugin {
 
         return min;
     }
+
+    /**
+     * Get the snapped to grid value.
```
Comment (thijstriemstra) [inline src/plugin/regions.js]: please expand this sentence, not clear to me what this does

File: src/plugin/regions.js (hunk 6 above)
Comment (moritzvieli) [reply src/plugin/regions.js]: I need access to it from the Region and from the RegionsPlugin. Is there a better place?

File: src/plugin/regions.js (hunk 5 above)
Comment (moritzvieli) [reply src/plugin/regions.js]: startTime has not been introduced with this PR. I tried to explain it better.

File: src/plugin/regions.js (hunk 4 above)
Comment (moritzvieli) [reply src/plugin/regions.js]: time has not been introduced with this PR. I tried to explain it better.

**Comment by thijstriemstra** [comment]
@moritzvieli this branch has conflicts that must be resolved first.

**Comment by moritzvieli** [comment]
@thijstriemstra The conflicts have been resolved.

**Comment by swampthang** [comment]
I'm hoping this is something that will allow me to set an incremental snap value so that when a region's handle is dragged or the entire region is dragged, it will snap to those values. However, I haven't found how to set this up. Can someone point me to or create an example of how to set this up? Really would appreciate it!

**Pull request merged**